chain/actors/agen/generator/gen.go

```diff
@@ -0,0 +1,223 @@
+package generator
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"text/template"
+
+	"golang.org/x/xerrors"
+)
+
+var latestVersion = 4
+
+var versions = []int{0, 2, 3, latestVersion}
+
+var versionImports = map[int]string{
+	0:             "/",
+	2:             "/v2/",
+	3:             "/v3/",
+	latestVersion: "/v4/",
+}
+
+var actors = map[string][]int{
+	"account":  versions,
+	"cron":     versions,
+	"init":     versions,
+	"market":   versions,
+	"miner":    versions,
+	"multisig": versions,
+	"paych":    versions,
+	"power":    versions,
+	"reward":   versions,
+	"verifreg": versions,
+}
+
+func Gen(pkg string) error {
+	if err := generateAdapters(pkg); err != nil {
+		return err
+	}
+
+	if err := generatePolicy(pkg); err != nil {
+		return err
+	}
+
+	if err := generateBuiltin(pkg); err != nil {
+		return err
+	}
+	return nil
+}
+
+func generateAdapters(pkg string) error {
+	actorsDir := "chain/actors/builtin"
+	for act, versions := range actors {
+		actDir := filepath.Join(actorsDir, act)
+
+		if err := generateState(pkg, actDir); err != nil {
+			return err
+		}
+
+		if err := generateMessages(pkg, actDir); err != nil {
+			return err
+		}
+
+		{
+			af, err := ioutil.ReadFile(filepath.Join(actDir, "actor.go.template"))
+			if err != nil {
+				return xerrors.Errorf("loading actor template: %w", err)
+			}
+
+			tpl := template.Must(template.New("").Funcs(template.FuncMap{
+				"import": func(v int) string { return versionImports[v] },
+			}).Parse(string(af)))
+
+			var b bytes.Buffer
+
+			err = tpl.Execute(&b, map[string]interface{}{
+				"versions":      versions,
+				"latestVersion": latestVersion,
+				"pkg":           pkg,
```

Comment:
The changes to this code lie here and elsewhere in template execution. Here is where it is used outside of louts: https://github.com/filecoin-project/sentinel-visor/pull/506/files#diff-4c7a9f71c8fc6368f8c09be8b31ccabdc897686acbc2d79734f694769e46750cR10-R11